File: src/engines/Greedy.java
package engines;

import game.Board;
import game.Move;
import game.Piece;
import utils.MoveGenerator;

import java.util.HashMap;
import java.util.List;

public class Greedy implements Engine {
    private final String name = "Greedy";
    private final String color;
    private final int depth;

    private final int alpha = Integer.MIN_VALUE;
    private final int beta = Integer.MAX_VALUE;

    private HashMap<Integer, Integer> pieceValues;

    public Greedy(String color) {
        this.color = color;
        initializePieceValues();
        this.depth = 4;
    }

    @Override
    public Move determineMove(Board board) {
        Move bestMove = new Move();
        int bestMoveEval = Integer.MIN_VALUE;
        int perspective = color.equals("w") ? 1 : -1;

        List<Move> moves = MoveGenerator.generateMoves(board);
        if (moves.size() == 0) {
            return null;
        }

        int count = 0;
        for (Move move : moves) {
            Board boardCopy = board.deepCopy();
            boardCopy.makeMove(move);
            boolean maximizingPlayer = boardCopy.getTurn() == Piece.WHITE;
            int[] moveEval = miniMax(boardCopy, this.depth, maximizingPlayer);
            count += moveEval[1];
            boardCopy = board.deepCopy();

            if (moveEval[0] * perspective > bestMoveEval) {
                bestMove = move;
                bestMoveEval = moveEval[0] * perspective;
            }
        }
        System.out.println(count);
        return bestMove;
    }

    private int[] miniMax(Board board, int depth, boolean maximizingPlayer) {
        if (depth == 0) {
            return new int[]{evaluate(board), 1};
        }

        List<Move> moves = MoveGenerator.generateMoves(board);
        int bestEval = (maximizingPlayer) ? Integer.MIN_VALUE : Integer.MAX_VALUE;
        int count = 0;

        for (Move move : moves) {
                Board boardCopy = board.deepCopy();
                boardCopy.makeMove(move);
                int[] moveEval = miniMax(boardCopy, depth - 1, !maximizingPlayer);
                count += moveEval[1];
                bestEval = (maximizingPlayer) ? Math.max(bestEval, moveEval[0]) : Math.min(bestEval, moveEval[0]);
        }
        return new int[]{bestEval, count};
    }

    private int evaluate(Board board) {
        int whiteEval = 0;
        int blackEval = 0;

        for (int index = 0; index < 64; index++) {
            if (board.isEmpty(index) || board.isKing(index)) {
                continue;
            }
            int value = pieceValues.get(board.getPieceType(index));

            if (board.isColor(index, Piece.WHITE)) {
                whiteEval += value;
            } else {
                blackEval += value;
            }

        }
        return whiteEval - blackEval;
    }

    private void initializePieceValues() {
        this.pieceValues = new HashMap<>();
        this.pieceValues.put(Piece.PAWN, 100);
        this.pieceValues.put(Piece.BISHOP, 300);
        this.pieceValues.put(Piece.KNIGHT, 300);
        this.pieceValues.put(Piece.ROOK, 500);
        this.pieceValues.put(Piece.QUEEN, 900);
    }

    @Override
    public String toString() {
        return "Greedy";
    }

    @Override
    public boolean isWhite() {
        return color.equals("w");
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public String getResults() {
        return null;
    }
}

File: src/engines/greedy/greedy.cpp
#include "greedy.h"

#include <bit>
#include <chrono>
#include <functional>
#include <iostream>

#include "../../moveGenerator.h"

Greedy::Greedy() {

}

std::string Greedy::name() const {
    return "Greedy";
}

Move Greedy::bestMove(Board& board, int time /*timeMs*/) {
    // stopRequested_.store(false, std::memory_order_relaxed);
    const bool maximizing = (board.getTurn() == Piece::WHITE);
    int nodes = 0;
    const auto t0 = std::chrono::steady_clock::now();

    Move best{};
    int bestScore = maximizing ? std::numeric_limits<int>::min() : std::numeric_limits<int>::max();;

    auto moves = MoveGenerator::generateMoves(board);
    if (moves.empty()) {
        return best;
    }

    for (const Move& m : moves) {
        // if (stopRequested_.load(std::memory_order_relaxed)) break;

        board.makeMove(m);
        int s = minimax(board, this->depth - 1, !maximizing, std::numeric_limits<int>::min(), std::numeric_limits<int>::max(), nodes);
        board.undoMove(m);

        if ((maximizing && s > bestScore) || (!maximizing && s < bestScore)) {
            bestScore = s;
            best = m;
        }
    }

    const auto t1 = std::chrono::steady_clock::now();
    auto us = std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count();
    double sec = us / 1e6;

    // std::cout << "[Minimax] depth=" << this->depth
    //           << " nodes=" << nodes
    //           << " time="  << std::fixed << std::setprecision(3) << (us / 1000.0) << " ms"
    //           << " nps="   << (sec > 0 ? static_cast<double>(nodes) / sec : 0.0)
    //           << " bestScore=" << bestScore
    //           << "\n";

    return best;
}

int Greedy::minimax(Board& board, int depth, bool maximizing, int alpha, int beta, int& nodes) {
    nodes++;

    // if (stopRequested_.load(std::memory_order_relaxed) || depth == 0 || board.isGameOver()) {
    if (depth == 0 || board.isGameOver()) {
        return evaluate(board);
    }

    auto moves = MoveGenerator::generateMoves(board);
    if (moves.empty()) {
        return evaluate(board);
    }

    if (maximizing) {
        int best = std::numeric_limits<int>::min();

        for (const Move& m : moves) {
            // if (stopRequested_.load(std::memory_order_relaxed)) break;

            board.makeMove(m);
            best = std::max(best, minimax(board, depth - 1, false, alpha, beta, nodes));
            board.undoMove(m);

            alpha = std::max(alpha, best);
            if (beta <= alpha) {
                break;
            }
        }

        return best;
    } else {
        int best = std::numeric_limits<int>::max();

        for (const Move& m : moves) {
            // if (stopRequested_.load(std::memory_order_relaxed)) break;

            board.makeMove(m);
            best = std::min(best, minimax(board, depth - 1, true, alpha, beta, nodes));
            board.undoMove(m);

            beta = std::min(beta, best);
            if (beta <= alpha) {
                break;
            }
        }

        return best;
    }
}

int Greedy::evaluate(Board& board) {
    if (board.isGameOver()) {
        if (board.isCheck() && board.getTurn() == Piece::WHITE) return std::numeric_limits<int>::max();
        if (board.isCheck() && board.getTurn() == Piece::BLACK) return std::numeric_limits<int>::min();

        //if it's not check, it's stalemate
        return 0;
    }

    int whiteScore = 0;
    int blackScore = 0;

    whiteScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::PAWN, Piece::WHITE)) * 100));
    whiteScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::KNIGHT, Piece::WHITE)) * 300));
    whiteScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::BISHOP, Piece::WHITE)) * 300));
    whiteScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::ROOK, Piece::WHITE)) * 500));
    whiteScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::QUEEN, Piece::WHITE)) * 900));

    blackScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::PAWN, Piece::BLACK)) * 100));
    blackScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::KNIGHT, Piece::BLACK)) * 300));
    blackScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::BISHOP, Piece::BLACK)) * 300));
    blackScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::ROOK, Piece::BLACK)) * 500));
    blackScore += (std::popcount(board.getPiecePositions(Piece::create(Piece::QUEEN, Piece::BLACK)) * 900));

    return whiteScore - blackScore;
}

void Greedy::stop() {
    stopRequested_.store(true, std::memory_order_relaxed);
}
